### PR TITLE
change(state): Set iterator read bounds where possible in DiskDb

### LINKED ISSUE #7664: Set iterator read bounds on deleting column families to avoid very slow syncs
### Motivation

Iterating on a column family that deletes lots of keys can be very slow.

We fixed some of this code in PRs #7663 and #7392, but we need a general fix to stop this issue happening again.

This currently impacts the `utxo_loc_by_transparent_addr_loc` column family, and any future uses of iterators with `zs_delete()`. It can be remotely triggered.

### Specifications

The RocksDB documentation recommends specifying lower and upper bounds on iterators:
https://github.com/facebook/rocksdb/wiki/Iterator#iterating-upper-bound-and-lower-bound
https://docs.rs/rocksdb/latest/rocksdb/struct.ReadOptions.html#method.set_iterate_range
https://docs.rs/rocksdb/latest/rocksdb/struct.DBCommon.html#method.iterator_cf_opt

This is how these bugs in other projects...
https://tracker.ceph.com/issues/55324
https://jira.mariadb.org/browse/MDEV-19670

Were fixed in those projects:
https://github.com/ceph/ceph/pull/46096/files
https://github.com/facebook/rocksdb/pull/5403

### Complex Code or Requirements

We need to use iterators for some operations, so our alternatives are (in preferred order):
1. Minimise the number of keys we delete, and how often we delete them (currently only UTXOs require key deletion)
2. Avoid using iterators on column families where we delete keys (currently only one column family does this, and it won't be necessary after PR #7392)
3. If we must use iterators on those column families, set read bounds to minimise the amount of deleted data that is read  (this ticket)

If this solution passes our benchmark tests, we should document it in:
https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/state-db-upgrades.md

### Testing

When we implement the test in ticket #7649, it will also cover this change.

### Documentation

We should document this issue in the developer state upgrade docs.

### Related Work

This is a long-term fix for #7618.

PR #7663 implements alternative 2 for some column families.
PR #7392 implements alternative 1 for some column families.

### PR DESCRIPTION
**Pull request by arya2**
## Motivation

Unbounded RocksDB iterators are slow in column families with many delete tombstones.

This PR updates `DiskDb` to bound iterators when reading a range of keys.

Close #7664.

## Solution

- Set lower and upper bounds on rocksdb iterators where possible in `DiskDb` methods

## Review

Anyone can review.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

- Refactor `address_utxo_locations()`

### CODE REVIEW

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (arya2) [inline zebra-state/src/service/finalized_state/disk_db.rs]: This is already tested by `all_upgrades_and_wrong_commitments_with_fake_activation_heights` using `zs_next_key_value_from()`.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db.rs:413]: This new argument needs an update to the method docs.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db.rs]: This code is technically correct, but might not achieve the performance goals, because the upper bound is larger than the minimum required upper bound.

I'd suggest using a well-tested big integer library function to do this.

Please also add a test case for the minimal upper bound. For example, the minimal upper bound of `0x00ff` is `0x0100`, not `0x01ff`.

**Review by teor2345 — DISMISSED** [review]
This refactor is nice and simple, but unfortunately the upper bound calculation is not minimal.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (arya2) [reply zebra-state/src/service/finalized_state/disk_db.rs]: Fixed in https://github.com/ZcashFoundation/zebra/pull/7731/commits/a703d1ce996256b02f13d707d51cde17d8d33c64.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (arya2) [reply zebra-state/src/service/finalized_state/disk_db.rs:413]: Updated in https://github.com/ZcashFoundation/zebra/pull/7731/commits/a703d1ce996256b02f13d707d51cde17d8d33c64.

File: zebra-state/src/service/finalized_state/disk_db/tests.rs
Comment (arya2) [inline zebra-state/src/service/finalized_state/disk_db/tests.rs]:
```suggestion
```

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_db.rs]: But unfortunately that test does not have sufficient coverage for the goals of the ticket.

File: zebra-state/src/service/finalized_state/disk_db/tests.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db/tests.rs]: This is missing a test case for u32::MAX

Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_db/tests.rs]: Otherwise this test has good coverage and maybe we don't need a proptest, since we know the structure of the code.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db.rs]: What happens if the key is variable-length?

**Review by teor2345 — COMMENTED** [review]
This looks great!

But there's one edge case we still need to cover: variable-length keys.

We don't have any at the moment, but PR #7392 changes the size of a key, and its validation runs an iterator with one key size on a column family which has been changed to a different key size.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (arya2) [reply zebra-state/src/service/finalized_state/disk_db.rs]: It would sometimes (probably <1%) fail to put an upper bound on iterators, which is still constrained by `take_while` but might unnecessarily read-ahead.

Its not an issue for the sprout/history tree change where the key size changes across db format versions, but if key structs had variable length serialized representations, like if heights 0-255 were just 1 byte, this would need something like a method in `IntoDisk` for checking the max serialization size of a struct.

Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_db.rs]: Zebra's data access patterns aren't random, so I'm not sure if that `<1%` probability will hold under deliberate attack, or with certain key formats. Here are two examples:

#### Attackers using maximum keys

In an attack or overload scenario, if the keys depend on mempool or block data, then many of them could be at the maximum value.

But this is unlikely to happen with the current UTXO key format, because the height comes first, and it is always less than the maximum. So if we decide to accept this risk, we should document it. (But see my alternative solution below.)

#### Zebra using maximum keys

In the sprout and history tree changes, the new key is `()`, which has an empty array as its serialization. This is the maximum value for an empty array. 

An earlier PR fixed this issue by using `zs_get()` to access keys, rather than an iterator.

But without that fix, every iterator that used `()` would read some tombstones from the previous height-based key format. This happens because the byte array increment code always returns `None` for `[0; u8]`. The missing bound would temporarily reduce performance after the format change, until those tombstones were compacted. (Sometimes that doesn't happen for a long time.)

#### Proposed Solution

If the key is the maximum for its current length, add an extra `0x01` byte to the front of the array. This will give you the  minimal bound.

Comment (arya2) [reply zebra-state/src/service/finalized_state/disk_db.rs]: > If the key is the maximum for its current length, add an extra 0x01 byte to the front of the array. This will give you the minimal bound.

Good idea.

Applied in https://github.com/ZcashFoundation/zebra/pull/7731/commits/fcf171f10be906f0c77041401ad1b59d5adadbaf.

File: zebra-state/src/service/finalized_state/disk_db/tests.rs
Comment (arya2) [reply zebra-state/src/service/finalized_state/disk_db/tests.rs]: > This is missing a test case for u32::MAX

Added in https://github.com/ZcashFoundation/zebra/pull/7731/commits/600aa5f6087dc0b6af9d9fcce54b614cedaee5c4.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (arya2) [inline zebra-state/src/service/finalized_state/disk_db.rs]:
```suggestion
                    *bound.get_mut(0).expect("should have at least 1 element") = 1;
```

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db.rs:325]: Nitpick: This IntoDisk bound is no longer needed. (Same for all the methods that use seek without using a key.)

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db.rs]: This is a substantial change which could change performance. Can you do a full sync before we merge this PR? Or split this change into a separate PR?

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db.rs]: Nitpick: potentially confusing documentation.

```suggestion
    /// Accepts a `reverse` argument. If it is `true`, creates the iterator with an [`IteratorMode`](rocksdb::IteratorMode)
```

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db.rs]: Nitpick: The `IntoDisk` bound is not needed here.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db.rs]: Optional Nitpick: RocksDB's documentation refers to keys as "lexicographically ordered bytes" rather than "big-endian numbers":
https://docs.rs/rocksdb/latest/rocksdb/struct.DBRawIteratorWithThreadMode.html#method.seek

```suggestion
                // and clear any bytes after it to increment the next key in lexicographic order (next big-endian number) this
```

File: zebra-state/src/service/finalized_state/disk_db/tests.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db/tests.rs:71]: Please check the "no added bytes" case here as well.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_db.rs]: Does each of the methods that calls `zs_seek()` have enough test coverage to make sure it works correctly? Or do we need to add `Some(_)` and `None` return value tests for the methods that aren't called from production code?

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_db.rs]: Can't we just push 0x01 in the line before?

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db.rs]: Optional Nitpick: "is_zero" is technically correct, but it makes the code hard to read, because the bound is actually beyond the maximum value for the original length of the key.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (arya2) [reply zebra-state/src/service/finalized_state/disk_db.rs]: > Can't we just push 0x01 in the line before?

I think that would put the 1 in the last element and that we need it in the first element.

Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_db.rs]: `bound.insert(0, 0x01)` achieves the same thing as these 2 lines, but it's much less verbose.

File: zebra-state/src/service/finalized_state/disk_db.rs
Comment (arya2) [reply zebra-state/src/service/finalized_state/disk_db.rs]: Split out to #7759.

**Review by teor2345 — DISMISSED** [review]
I'm happy to merge if these extra checks pass

File: zebra-state/src/service/finalized_state/disk_db/tests.rs
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_db/tests.rs:71]: I added this check

File: zebra-state/src/service/finalized_state/disk_db/tests.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db/tests.rs]:
```suggestion
                key.to_be_bytes().len(),
```

File: zebra-state/src/service/finalized_state/disk_db/tests.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_db/tests.rs]:
```suggestion
            bytes[0], 0x00,
```

**Review by teor2345 — DISMISSED** [review]

**Review by teor2345 — DISMISSED** [review]

**Comment by gustavovalverde** [comment]
@mergifyio refresh

**Review by teor2345 — APPROVED** [review]
Thanks!

**Pull request merged**
